tools/wptrunner/wptrunner/executors/base.py

```diff
@@ -476,3 +476,83 @@ def is_alive(self):
         conn.request("HEAD", self.server.base_path + "invalid")
         res = conn.getresponse()
         return res.status == 404
+
+
+class CallbackHandler(object):
+    """Handle callbacks from testdriver-using tests.
+
+    The default implementation here makes sense for things that are roughly like
+    WebDriver. Things that are more different to WebDriver may need to create a
+    fully custom implementation."""
+
+    def __init__(self, logger, protocol, test_window):
+        self.protocol = protocol
+        self.test_window = test_window
+        self.logger = logger
+        self.callbacks = {
+            "action": self.process_action,
+            "complete": self.process_complete
+        }
+
+        self.actions = {
+            "click": ClickAction(self.logger, self.protocol)
+        }
+
+    def __call__(self, result):
+        self.logger.debug("Got async callback: %s" % result[1])
+        try:
+            return self.callbacks[result[1]](result)
+        except KeyError:
+            raise ValueError("Unknown callback type %r" % result[1])
+
+    def process_complete(self, result):
+        rv = [result[0]] + result[2]
```

Comment:
I think it’s worth documenting the message format somewhere. Reverse engineering the test data format when I implemented the pytest runner was not my greatest fun-packed day at work.